### ISSUE TITLE
mpsc channel should have an is_closed function

### ISSUE THREAD

**Issue by estokes** [tokio-rs/tokio#2469]
Currently it can be done using `poll_ready` and `disarm`:
``` rust
async fn channel_closed<T>(channel: &Sender<T>) -> bool {
    match poll_fn(|cx| Poll::Ready(chan.poll_ready(cx))).await {
        Poll::Ready(Ok(())) => { chan.disarm(); false },
        Poll::Ready(Err(_)) => true,
        Poll::Pending => false,
    }
}
```

**Comment by estokes** [comment]
It's a bit of a bother if this has to be an async fn, for example it'd be nice to use it in e.g. vec.retain(|c| !c.is_closed()), and similar situations.

**Comment by zeroed** [comment]
Hello, I would like to work on this issue.

I've made a very exploratory attempt ([here a branch](https://github.com/tokio-rs/tokio/compare/master...zeroed:mpsc-channel-is-closed)) which by no means I would consider concluded, but if I may, I would like to ask for an early feedback to validate my approach.

I'm not yet completely comfortable with the internals of `sync::mpsc`; according to my understanding a `Sender` wraps a `chan::Tx` and a `Receiver` wraps a `chan::Rx`. Both `Tx` and `Rx` have an `inner inner: Arc<Chan>`. A `Chan` has a `tx_count` that is incremented for every `clone` of `Tx` and decremented on `drop`, and it has a `rx_fields` with a boolean field `rx_closed`.

In my approach I would use the `tx_count` and `rx_closed` fields to implement a function `is_closed` in both `Tx` and `Rx` (then have it exposed in `Sender` and `Receiver`, bounded and unbounded).

A very simple tests seems to support the idea but I may have overlooked something and I would like your feedback about this. Lastly, I'm not entirely convinced by the partial repetition in the 2 `is_close` implementations.

Proper documentation will be surely added.

Thank you in advance.

**Comment by estokes** [comment]
I'd like to add to this discussion that the futures-rs mpsc channel's ability to tell you whether two senders are from the same receiver, as well as hashing a sender based on the receiver they go to is a real super power sometimes. It would be really nice to also have this functionality in tokio. The two functions in question are,

```rust
pub fn same_receiver(&self, other: &Sender<T>) -> bool
pub fn hash_receiver<H>(&self, hasher: &mut H)
```

**Comment by Darksonn** [comment]
Those should go in a new issue so we can close this issue once the PR is merged.

**Comment by LucioFranco** [comment]
Bumping this: I would like to prioritize this.

**Comment by MikailBag** [comment]
It seems that `async`-ness can be avoided pretty easily, constructing fake waker:
```rust
fn channel_closed<T>(chan: &mut Sender<T>) -> bool {
    let mut cx = Context::from_waker(noop_waker_ref());
    match chan.poll_ready(&mut cx) {
        Poll::Ready(Ok(())) => { chan.disarm(); false },
        Poll::Ready(Err(_)) => true,
        Poll::Pending => false,
    }
}
```
playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3ee44016bbb96f5bf054ccd15246182e

**Comment by Darksonn** [comment]
If some other task is also polling the channel, then that task would no longer receive wake-ups. It would also be nice to do the check without having to disarm afterwards.

**Comment by LucioFranco** [comment]
> If some other task is also polling the channel, then that task would no longer receive wake-ups

I am not directly following what you mean by this?

**Comment by Darksonn** [comment]
If I call `poll_ready` with some waker, and then call `is_closed`, my waker is no longer scheduled for wake-ups. This would be surprising. Of course, if we remove poll methods shortly, the issue goes away.
